flocker/provision/_aws.py

```diff
@@ -27,70 +34,232 @@
 }
 
 
-def get_default_username(distribution):
+IMAGE_NAMES = {
+    'centos-7': 'CentOS 7 x86_64 (2014_09_29) EBS HVM'
+                '-b7ee8a69-ee97-4a49-9e68-afaee216db2e-ami-d2a117ba.2',
+    'ubuntu-14.04': 'ubuntu/images/hvm-ssd/ubuntu-trusty-14.04-amd64-server-20150325',  # noqa
+    'ubuntu-15.04': 'ubuntu/images/hvm-ssd/ubuntu-vivid-15.04-amd64-server-20150422',  # noqa
+}
+
+
+def _wait_until_running(instance):
     """
-    Return the username available by default on a system.
+    Wait until a instance is running.
 
-    :param str distribution: Name of the operating system distribution
-    :return str: The username made available by AWS for this distribution.
+    :param boto.ec2.instance.Instance instance: The instance to wait for.
     """
-    return _usernames[distribution]
+    with start_action(
+        action_type=u"flocker:provision:aws:wait_until_running",
+        instance_id=instance.id,
+    ):
+        state = instance.state
+        while state != 'running':
+            with start_action(
+                action_type=u"flocker:provision:aws:wait_until_running:sleep",
+                instance_state=instance.state,
+            ):
+                sleep(1)
+            instance.update()
+            new_state = instance.state
+            if new_state != state:
+                state = new_state
+
+
+@implementer(INode)
+class AWSNode(PClass):
+    _provisioner = field(mandatory=True)
+    _instance = field(mandatory=True)
+    distribution = field(mandatory=True)
+    name = field(mandatory=True)
+
+    @property
+    def address(self):
+        return self._instance.ip_address.encode('ascii')
+
+    @property
+    def private_address(self):
+        return self._instance.private_ip_address.encode('ascii')
+
+    def destroy(self):
+        with start_action(
+            action_type=u"flocker:provision:aws:destroy",
+            instance_id=self._instance.id,
+        ):
+            self._instance.terminate()
+
+    def get_default_username(self):
+        """
+        Return the username available by default on a system.
+        """
+        return _usernames[self.distribution]
+
+    def provision(self, package_source, variants=()):
+        """
+        Provision flocker on this node.
+
+        :param LibcloudNode node: Node to provision.
+        :param PackageSource package_source: See func:`task_install_flocker`
+        :param set variants: The set of variant configurations to use when
+            provisioning
+        """
+        username = self.get_default_username()
+
+        commands = []
 
+        # cloud-init may not have allowed sudo without tty yet, so try SSH key
+        # installation for a few more seconds:
+        start = []
 
-def provision_aws(node, package_source, distribution, variants):
+        def for_ten_seconds(*args, **kwargs):
+            if not start:
+                start.append(time())
+            return Effect(Constant((time() - start[0]) < 30))
```

Comment:
This seems to wait 30 seconds rather than 10?  Definitely change the function name. Maybe a better solution is something like:
```
def for_time_period(seconds):
    start = []
    def check(exc_info):
        if not start:
            start.append(time())
        return Effect(Constant((time() - start[0]) < seconds))
    return check

... commands=retry(task_install_ssh_key(), for_time_period(30))
```
This also sets us up for checking any permanent exceptions that should not be retried.